tests/test_api.py

```diff
@@ -1055,3 +1056,38 @@ def test_open_array_with_mode_r_plus(store: Store) -> None:
     assert isinstance(z2, Array)
     assert (z2[:] == 1).all()
     z2[:] = 3
+
+
+@pytest.mark.parametrize("func_name", zarr.api.synchronous.__all__)
+def test_derived_docstrings(func_name: str) -> None:
+    """
+    Test that functions in the synchronous API module have
+    docstrings that are derived from those in the asynchronous API module
+    """
+    if func_name in ("open_array",):
+        msg = (
+            '"open_array" gets xfailed because the async version and the sync '
+            "version have meaningful differences in their docstrings."
+        )
+        pytest.xfail(reason=msg)
+    assert (
+        getattr(zarr.api.synchronous, func_name).__doc__
+        == getattr(zarr.api.asynchronous, func_name).__doc__
+    )
+
+
+@pytest.mark.parametrize("func_name", zarr.api.synchronous.__all__)
+def test_derived_signatures(func_name: str) -> None:
+    """
+    Test that functions in the API module have signatures that are derived from those in the asynchronous API module
+    """
+    if func_name in ("tree", "load", "open", "open_array", "group"):
```

Comment:
If the signature of two functions is different, shouldn't the docstring be different too. ie., this list and the same list above in `test_derived_docstrings` should be the same?

In at least `tree()` this seems to be correct - the docstring for `zarr.api.asynchronous.tree` needs correcting to say it takes an `AsyncGroup` as input, at which point it will have a different docstring to `zarr.api.synchronous.tree`

Reply:
yes i think you are right -- by construction _all_ of the docstrings should be different in their return type annotation, and so this test is currently incorrect (fortunately I guarded for something like this outcome in the test docstring by stating "derived from" rather than "equal to")